<a href="https://colab.research.google.com/github/OneeyedCheshire/COursera-Capstone/blob/master/IBM_Data_Science_Capstone_Week_3_Toronto_Segmentation_Jason_Meil_ipynb_txt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Install packages:
1. beautySoup4 to scrape Wiki page
pip install beautysoup5

2. The xml parser to parse page:
pip install lxmp

pip install html5lib

pip install requests

First import all libraries.

In [0]:
from bs4 import BeautifulSoup
import requests
import lxml.html as lh
import pandas as pd

In [2]:

url = 'http://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'


page = requests.get(url)


doc = lh.fromstring(page.content)



tr_elements = doc.xpath('//tr')


[len(T) for T in tr_elements[:12]]


[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

In [3]:
tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print '%d:"%s"'%(i,name)
    col.append((name,[]))

1:"Postcode"
2:"Borough"
3:"Neighbourhood
"


In [0]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

Just to be sure, let’s check the length of each column. Ideally, they should all be the same.

In [0]:
[len(C) for (title,C) in col]

[289, 289, 289]

CREATING  a DataFrame

In [5]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)
df.columns=['Borough',"Neighbourhood","Postcode"]
print(df.head())
print(df.shape)

            Borough       Neighbourhood Postcode
0      Not assigned      Not assigned\n      M1A
1      Not assigned      Not assigned\n      M2A
2        North York         Parkwoods\n      M3A
3        North York  Victoria Village\n      M4A
4  Downtown Toronto      Harbourfront\n      M5A
(289, 3)


## Cleaning our data

In [7]:
# 1. removing all occurence of  \n
df=df.replace({'\n': ''}, regex=True)
print(df.head())

            Borough     Neighbourhood Postcode
0      Not assigned      Not assigned      M1A
1      Not assigned      Not assigned      M2A
2        North York         Parkwoods      M3A
3        North York  Victoria Village      M4A
4  Downtown Toronto      Harbourfront      M5A


In [0]:
#2. Dropping  rows with a borough that is Not assigned.
print(df[df['Borough']=="Not assigned"].shape)
df=df[(df['Borough']!="Not assigned")]
df.head()

(77, 3)


Borough     Neighbourhood Postcode
2        North York         Parkwoods      M3A
3        North York  Victoria Village      M4A
4  Downtown Toronto      Harbourfront      M5A
5  Downtown Toronto       Regent Park      M5A
6        North York  Lawrence Heights      M6A

We have 77 rows with Borough that is Not assigned.

In [0]:
# Check unique values in each column:
print(len(df.Borough.unique()))
print(len(df.Postcode.unique()))
print(len(df.Neighbourhood.unique()))

11
103
210


We have 11 unique Borough, 103 unique Postcodes, 209 unique Neighbourhood. 


## Processing Data Frame
#4. Combining multiple neighborhood values into one rows for one postal code area.

In [12]:
df_test = df
df_g=df_test.groupby('Postcode')['Neighbourhood'].apply(','.join).reset_index()
df_g


Postcode                                      Neighbourhood
0        M1A                                       Not assigned
1        M1B                                      Rouge,Malvern
2        M1C               Highland Creek,Rouge Hill,Port Union
3        M1E                    Guildwood,Morningside,West Hill
4        M1G                                             Woburn
5        M1H                                          Cedarbrae
6        M1J                                Scarborough Village
7        M1K          East Birchmount Park,Ionview,Kennedy Park
8        M1L                      Clairlea,Golden Mile,Oakridge
9        M1M      Cliffcrest,Cliffside,Scarborough Village West
10       M1N                         Birch Cliff,Cliffside West
11       M1P  Dorset Park,Scarborough Town Centre,Wexford He...
12       M1R                                   Maryvale,Wexford
13       M1S                                          Agincourt
14       M1T              Clarks Corners,Sullivan,Tam O'Shanter
15       M1V  Agincourt North,L'Amoreaux East,Milliken,Steel...
16       M1W                       L'Amoreaux West,Steeles West
17       M1X                                        Upper Rouge
18       M1Y                                       Not assigned
19       M1Z                                       Not assigned
20       M2A                                       Not assigned
21       M2B                                       Not assigned
22       M2C                                       Not assigned
23       M2E                                       Not assigned
24       M2G                                       Not assigned
25       M2H                                  Hillcrest Village
26       M2J                         Fairview,Henry Farm,Oriole
27       M2K                                    Bayview Village
28       M2L                            Silver Hills,York Mills
29       M2M                             Newtonbrook,Willowdale
..       ...                                                ...
150      M8N                                       Not assigned
151      M8P                                       Not assigned
152      M8R                                       Not assigned
153      M8S                                       Not assigned
154      M8T                                       Not assigned
155      M8V         Humber Bay Shores,Mimico South,New Toronto
156      M8W                              Alderwood,Long Branch
157      M8X        The Kingsway,Montgomery Road,Old Mill North
158      M8Y  Humber Bay,King's Mill Park,Kingsway Park Sout...
159      M8Z  Kingsway Park South West,Mimico NW,The Queensw...
160      M9A                                   Islington Avenue
161      M9B  Cloverdale,Islington,Martin Grove,Princess Gar...
162      M9C  Bloordale Gardens,Eringate,Markland Wood,Old B...
163      M9E                                       Not assigned
164      M9G                                       Not assigned
165      M9H                                       Not assigned
166      M9J                                       Not assigned
167      M9K                                       Not assigned
168      M9L                                      Humber Summit
169      M9M                                    Emery,Humberlea
170      M9N                                             Weston
171      M9P                                          Westmount
172      M9R  Kingsview Village,Martin Grove Gardens,Richvie...
173      M9S                                       Not assigned
174      M9T                                       Not assigned
175      M9V  Albion Gardens,Beaumond Heights,Humbergate,Jam...
176      M9W                                          Northwest
177      M9X                                       Not assigned
178      M9Y                                       Not assigned
179      M9Z                                       Not assigned

[180 rows x 2 columns]

In [13]:
df_part2=df[["Borough","Postcode"]] 
df_part2.head()

Borough Postcode
0      Not assigned      M1A
1      Not assigned      M2A
2        North York      M3A
3        North York      M4A
4  Downtown Toronto      M5A

In [0]:
#Joining two dataframes 
df_res=df_g.join(df_part2.set_index('Postcode'), on='Postcode')


In [15]:
#removing duplicates:
df_res_unique = df_res.drop_duplicates(subset=['Postcode', 'Neighbourhood','Borough'], keep=False)
df_res_unique.shape

(122, 3)

In [0]:
df_test = df_res_unique

In [17]:
df_test[df_test['Neighbourhood']=="Not assigned"]

Postcode Neighbourhood       Borough
0        M1A  Not assigned  Not assigned
18       M1Y  Not assigned  Not assigned
19       M1Z  Not assigned  Not assigned
20       M2A  Not assigned  Not assigned
21       M2B  Not assigned  Not assigned
22       M2C  Not assigned  Not assigned
23       M2E  Not assigned  Not assigned
24       M2G  Not assigned  Not assigned
33       M2S  Not assigned  Not assigned
34       M2T  Not assigned  Not assigned
35       M2V  Not assigned  Not assigned
36       M2W  Not assigned  Not assigned
37       M2X  Not assigned  Not assigned
38       M2Y  Not assigned  Not assigned
39       M2Z  Not assigned  Not assigned
43       M3E  Not assigned  Not assigned
44       M3G  Not assigned  Not assigned
51       M3P  Not assigned  Not assigned
52       M3R  Not assigned  Not assigned
53       M3S  Not assigned  Not assigned
54       M3T  Not assigned  Not assigned
55       M3V  Not assigned  Not assigned
56       M3W  Not assigned  Not assigned
57       M3X  Not assigned  Not assigned
58       M3Y  Not assigned  Not assigned
59       M3Z  Not assigned  Not assigned
79       M4Z  Not assigned  Not assigned
98       M5Y  Not assigned  Not assigned
99       M5Z  Not assigned  Not assigned
114      M6T  Not assigned  Not assigned
..       ...           ...           ...
134      M7T  Not assigned  Not assigned
135      M7V  Not assigned  Not assigned
136      M7W  Not assigned  Not assigned
137      M7X  Not assigned  Not assigned
139      M7Z  Not assigned  Not assigned
140      M8A  Not assigned  Not assigned
141      M8B  Not assigned  Not assigned
142      M8C  Not assigned  Not assigned
143      M8E  Not assigned  Not assigned
144      M8G  Not assigned  Not assigned
145      M8H  Not assigned  Not assigned
146      M8J  Not assigned  Not assigned
147      M8K  Not assigned  Not assigned
148      M8L  Not assigned  Not assigned
149      M8M  Not assigned  Not assigned
150      M8N  Not assigned  Not assigned
151      M8P  Not assigned  Not assigned
152      M8R  Not assigned  Not assigned
153      M8S  Not assigned  Not assigned
154      M8T  Not assigned  Not assigned
163      M9E  Not assigned  Not assigned
164      M9G  Not assigned  Not assigned
165      M9H  Not assigned  Not assigned
166      M9J  Not assigned  Not assigned
167      M9K  Not assigned  Not assigned
173      M9S  Not assigned  Not assigned
174      M9T  Not assigned  Not assigned
177      M9X  Not assigned  Not assigned
178      M9Y  Not assigned  Not assigned
179      M9Z  Not assigned  Not assigned

[78 rows x 3 columns]

In [18]:
#only one value, easely replace it
df_test=df_test.set_value(85, 'Neighbourhood', 'Queen\'s Park')

/usr/local/lib/python2.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: set_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead
  """Entry point for launching an IPython kernel.
/usr/local/lib/python2.7/dist-packages/pandas/core/frame.py:2035: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.loc[index, col] = value


In [19]:
df_test.head(10)

Postcode        Neighbourhood       Borough
0       M1A         Not assigned  Not assigned
4       M1G               Woburn   Scarborough
5       M1H            Cedarbrae   Scarborough
6       M1J  Scarborough Village   Scarborough
13      M1S            Agincourt   Scarborough
17      M1X          Upper Rouge   Scarborough
18      M1Y         Not assigned  Not assigned
19      M1Z         Not assigned  Not assigned
20      M2A         Not assigned  Not assigned
21      M2B         Not assigned  Not assigned

In [21]:
df.shape()

TypeError: ignored